src/awkward1/operations/convert.py

```diff
@@ -1139,6 +1139,180 @@ def getfunction(layout, depth):
     else:
         return out
 
+def to_arrow(layout):
+    import pyarrow
+
+    def recurse(layout, mask = None):
+
+        if isinstance(layout, awkward1.layout.NumpyArray):
+            numpy_arr = numpy.asarray(layout)
+            if(numpy_arr.ndim == 1):
+                return pyarrow.array(numpy_arr, mask=mask)
+            else:
+                return pyarrow.Tensor.from_numpy(numpy_arr)
```

Comment:
I have my doubts regarding entrusting the `multidimensional numpy arrays` to `pyarrow Tensors` since there is no straightforward way of accommodating for the lack of masks. In my opinion, maybe handling this with a `ListOffsetArray` should be a better choice, since you can specify fixed offsets in a way similar to `RegularArray`.

Reply:
Maybe we should draw Venn diagrams: Awkward has high-level ListType (ListArray and ListOffsetArray) and RegularType, which can contain anything; Arrow has one variable-length list type and Tensor, for the special case of regular arrays of numbers. If the Awkward structure in question is a NumpyArray with len(shape) > 1 or a NumpyArray nested within RegularArrays, then it can go to Tensor. Anything else can't and has to defer to Arrow's list type.

For Awkward RegularArrays (containing anything other than RegularArrays and NumpyArrays), that means generating regular offsets, as you say. For Awkward ListArrays, that means "compacting" the list because Arrow uses offsets, not starts and stops. Fortunately, all three of these classes have a `toListOffsetArray64` method that normalizes the data (to an Arrow LargeListArray because it's 64-bit, and we could add a `toListOffsetArray32` in the future if it's useful to make smaller Arrow buffers).

About masking, it's true that something like `RegularArray(BitMaskedArray(RegularArray(NumpyArray(...))))` can't be expressed as an Arrow Tensor because the option-type is in the middle. It represents data like `[None, [2, 3]]`, which is not something an Arrow Tensor can represent. One option is to push all the Nones to the leaf level, like `[[None, None], [2, 3]]`, which is what `ak.to_numpy` does (producing a NumPy MaskedArray). But it's also okay to punt and just say that only structures like `RegularArray(RegularArray(BitMaskedArray(NumpyArray(...))))` are allowed.

(Or am I wrong and Arrow's Tensor only takes masks at the outside, like `BitMaskedArray(RegularArray(RegularArray(NumpyArray(...))))`? If so, that would be strange because it's the opposite of NumPy.)